core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -451,16 +452,24 @@ final class ShuffleBlockFetcherIterator(
           var isStreamCopied: Boolean = false
           try {
             input = streamWrapper(blockId, in)
-            // Only copy the stream if it's wrapped by compression or encryption, also the size of
-            // block is small (the decompressed block is smaller than maxBytesInFlight)
-            if (detectCorrupt && !input.eq(in) && size < maxBytesInFlight / 3) {
+            // Only copy the stream if it's wrapped by compression or encryption upto a size of
+            // maxBytesInFlight/3. If stream is longer, then corruption will be caught while reading
+            // the stream.
+            streamCompressedOrEncrypted = !input.eq(in)
+            if (detectCorrupt && streamCompressedOrEncrypted) {
               isStreamCopied = true
               val out = new ChunkedByteBufferOutputStream(64 * 1024, ByteBuffer.allocate)
-              // Decompress the whole block at once to detect any corruption, which could increase
-              // the memory usage tne potential increase the chance of OOM.
+              // Decompress the block upto maxBytesInFlight/3 at once to detect any corruption which
+              // could increase the memory usage and potentially increase the chance of OOM.
               // TODO: manage the memory used here, and spill it into disk in case of OOM.
-              Utils.copyStream(input, out, closeStreams = true)
-              input = out.toChunkedByteBuffer.toInputStream(dispose = true)
+              isStreamCopied = Utils.copyStreamUpTo(
+                input, out, maxBytesInFlight / 3, closeStreams = true)
```

Comment:
What is the reasoning behind considering one-third of the block, why for example not one-fourth?

Reply:
That limit was already defined. I did not change that limit. The purpose was to limit the amount of memory used but I am not sure how we came up with that exact threshold.

Reply:
I can't find where it is defined. I thought your change is proposing that. Sorry if I missed that

Reply:
Look for this line on left side of the files changed: `if (detectCorrupt && !input.eq(in) && size < maxBytesInFlight / 3) {`